src/value.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright 2024 the Interpoli Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use alloc::vec::Vec;
+use peniko::{self, kurbo};
+
+/// Fixed or animated value.
+#[derive(Clone, Debug)]
+pub enum Value<T: Tween> {
+    /// Fixed value.
+    Fixed(T),
+    /// Animated value.
+    Animated(Animated<T>),
+}
+
+impl<T: Tween> Value<T> {
+    /// Returns true if the value is fixed.
+    pub fn is_fixed(&self) -> bool {
+        matches!(self, Self::Fixed(_))
+    }
+
+    /// Returns the value at a specified frame.
+    pub fn evaluate(&self, frame: f64) -> T {
+        match self {
+            Self::Fixed(fixed) => fixed.clone(),
+            Self::Animated(animated) => animated.evaluate(frame),
+        }
+    }
+}
+
+impl<T: Tween + Default> Default for Value<T> {
+    fn default() -> Self {
+        Self::Fixed(T::default())
+    }
+}
+
+/// Borrowed or owned value.
+#[derive(Clone, Debug)]
+pub enum ValueRef<'a, T> {
+    Borrowed(&'a T),
+    Owned(T),
+}
+
+impl<'a, T> AsRef<T> for ValueRef<'a, T> {
+    fn as_ref(&self) -> &T {
+        match self {
+            Self::Borrowed(value) => value,
+            Self::Owned(value) => value,
+        }
+    }
+}
+
+impl<'a, T: Clone> ValueRef<'a, T> {
+    pub fn into_owned(self) -> T {
+        match self {
+            Self::Borrowed(value) => value.clone(),
+            Self::Owned(value) => value,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Easing {
+    pub o: EasingHandle,
+    pub i: EasingHandle,
+}
+
+impl Easing {
+    pub const LERP: Easing = Easing {
+        o: EasingHandle { x: 0.0, y: 0.0 },
+        i: EasingHandle { x: 1.0, y: 1.0 },
+    };
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct EasingHandle {
+    pub x: f64,
+    pub y: f64,
+}
+
+/// Time for a particular keyframe, represented as a frame number.
+#[derive(Copy, Clone, Debug)]
+pub struct Time {
+    /// Frame number.
+    pub frame: f64,
+    /// Easing tangent going into the next keyframe
+    pub in_tangent: Option<EasingHandle>,
+    /// Easing tangent leaving the current keyframe
+    pub out_tangent: Option<EasingHandle>,
+    /// Whether it's a hold frame.
+    pub hold: bool,
+}
+
+impl Time {
+    /// Returns the frame indices and interpolation weight for the given frame,
+    /// and whether to hold the frame
+    pub(crate) fn frames_and_weight(
+        times: &[Time],
+        frame: f64,
+    ) -> Option<([usize; 2], f64, Easing, bool)> {
+        if times.is_empty() {
+            return None;
+        }
+        use core::cmp::Ordering::*;
+        let ix = match times.binary_search_by(|x| {
```

Comment:
Can this use [`binary_search_by_key`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search_by_key)

Reply:
If I recall (going off poor memory here) that didn't work because it's a float which doesn't have `Eq`. But, given that, I'd still rewrite this to `a.frame.partial_cmp(b).unwrap_or(Equal)`